### ISSUE TITLE
Simplify taking a snapshot/sampling with an IDE shortcut

### ISSUE THREAD

**Issue by JaroslavTulach** [enso-org/enso#6775]
The amount of non-reproducible issues is increasing. Having a video of the failure isn't usually enough to understand the problem. We need deeper insight into the engine. We have description of few [manual steps](https://github.com/enso-org/enso/issues/6672#issuecomment-1551344980) one can follow to gather _thread dump_, _sampling snapshot_ or _heap dump_. However these steps involve [VisualVM](https://www.graalvm.org/22.3/tools/visualvm/) and may be hard to follow for non-Java developers. We need to simplify the process of obtaining the data.

### UX

- user sees a problem with the engine
- user presses a shortcut
- detailed information about the engine process is stored into files next to log files
- a message where to find the files is shown to the user
- user can upload the files at google drive and link them from a bug report

### Increase Verbosity of Logging

The engine shall increase the verbosity of logging into a rotating `.log` files (located at [user data directory](https://github.com/enso-org/enso/blob/develop/docs/distribution/distribution.md#installed-enso-distribution-layout))  by default (suggested by @hubertp). The console logging may still contain just the most important log messages. 

- #6789

Issue reporting template shall require log files to be attached (suggested by @4e6) and the _snapshot_ (see next) shall contain the `.log` file for further analysis.

We should also make sure that the performance impact is negligible with very verbose logs 
- #6739

### Taking a Snapshot

- IDE captures the shortcut and sends LS/engine a message requesting to "take a snapshot"
- the JVM dumps its state into an `.hprof` file (includes heap dump as well as thread dump)
- sends IDE back a response with the location of the file
- IDE shows a message to the user with location of the file

### Self-Sampling

The engine has a capability of [self profiling](https://bits.netbeans.org/dev/javadoc/org-netbeans-modules-sampler/org/netbeans/modules/sampler/Sampler.html) at any time. Currently one can [enable this only at startup](https://github.com/enso-org/enso/blob/78c36ddfc79e943b272d46dd68dc6277af359c5d/docs/profiler/engine-startup.md?plain=1#L16) via `--profiling-path` & co. options. However we can expose this functionality via a shortcut as well. For example:

- user presses a shortcut - self sampling starts
- user performs a problematic action in the IDE
- user presses a shortcut again to stop the sampling
- two files `.npss` & `.log` are written next to log files
- a message with their location is shown to the user

Self-sampling could be useful for understanding performance issues as well as understanding what the engine is actually doing when handling the messages.

### Cloud!?

What to do with cloud, @PabloBuchu? Can the user get the log file? Are they stored somewhere? Is it possible to identify them with a session ID, if such session ID is attached to the bug report?

**Comment by hubertp** [comment]
One thing which I'd like us to be a bit more careful is to **not** log most of the suggestions updates by default. In verbose mode we are logging all documentation that is sent over the wire which is really annoying and hinders debugging attempts.

**Comment by JaroslavTulach** [comment]
Another example of usefulness of investing in better _"snapshot taking"_ infrastructure: If the bug report for 
- #6841 

contained _thread dump_, then it would be easy to realize it is caused by a *deadlock* without hours of attempting to reproduce the problem. Creating #6920 fix would be way easier.

**Comment by jdunkerley** [comment]
We want an integrated solution within the chat window allowing users to gather the logs and send them to us. This ticket to add a part to the project manager to gather the logs (as a zip or tar.gz). 

Will need the dashboard team to then gather from here and send to the chat thread.

**Comment by hubertp** [comment]
So both, local and cloud, should generate a single tarball that is easy to upload:
- for local, the LS endpoint could return the path on where the tarball is located
- for cloud, the LS endpoint could upload it to some predefined s3 bucket (LS would need to be started with some auth keys and the location defined in the config file)

**Comment by PabloBuchu** [comment]
| What to do with cloud, @PabloBuchu? Can the user get the log file? Are they stored somewhere? Is it possible to identify them with a session ID, if such session ID is attached to the bug report?

Every time user opens a project and start engine for their need we associate a `SessionId` with that particular run. We send this session id with IDE and cloud logs for easy tracking. We can also inject this to engine via env variable?

So this generated snapshot should be created in a default directory and the path to that directory should be managable via env variables. This way when run in docker container we can create shanpshot on persistent mounted volume. 

Creating an endpoint for uploading this tarballs is also simple from our side

**Comment by PabloBuchu** [comment]
| for cloud, the LS endpoint could upload it to some predefined s3 bucket (LS would need to be started with some auth keys and the location defined in the config file)

we already have an authentication system node -> API so this should be reused. LS needs I guess fetch auth_token, api gateway url and endpoint to send package to the cloud

**Comment by 4e6** [comment]
Backend sampling, memory snapshot, and profiling shortcuts were implemented by the linked PRs.
